cmd/tailscale/cli/update.go

```diff
@@ -200,6 +271,169 @@ func (up *updater) updateWindows() error {
 	if !winutil.IsCurrentProcessElevated() {
 		return errors.New("must be run as Administrator")
 	}
-	// TODO(bradfitz): require elevated mode
-	return errors.New("TODO: download + msiexec /i /quiet " + url)
+	if err := up.confirm(ver); err != nil {
+		return err
+	}
+
+	targetDir := filepath.Join(os.Getenv("ProgramData"), "Tailscale", "MSICache")
+	if err := os.MkdirAll(targetDir, 0700); err != nil {
+		return err
+	}
+	msiTarget := filepath.Join(targetDir, path.Base(url))
+	if err := downloadURLToFile(url, msiTarget); err != nil {
+		return err
+	}
+
+	log.Printf("copying tailscaled.exe to copy...")
+	selfCopy, err := makeSelfCopy()
+	if err != nil {
+		return err
+	}
+	defer os.Remove(selfCopy)
+	log.Printf("running copy of self...")
+
+	cmd := exec.Command(selfCopy, "update")
```

Comment:
msiexec handles signature checking? I suppose it would.

Reply:
Presumably!

The wireguard-windows MSI updater additionally does:

```go
func verifyAuthenticode(path string) bool {
	path16, err := windows.UTF16PtrFromString(path)
	if err != nil {
		return false
	}
	data := &windows.WinTrustData{
		Size:             uint32(unsafe.Sizeof(windows.WinTrustData{})),
		UIChoice:         windows.WTD_UI_NONE,
		RevocationChecks: windows.WTD_REVOKE_WHOLECHAIN, // Full revocation checking, as this is called with network connectivity.
		UnionChoice:      windows.WTD_CHOICE_FILE,
		StateAction:      windows.WTD_STATEACTION_VERIFY,
		FileOrCatalogOrBlobOrSgnrOrCert: unsafe.Pointer(&windows.WinTrustFileInfo{
			Size:     uint32(unsafe.Sizeof(windows.WinTrustFileInfo{})),
			FilePath: path16,
		}),
	}
	verified := windows.WinVerifyTrustEx(windows.InvalidHWND, &windows.WINTRUST_ACTION_GENERIC_VERIFY_V2, data) == nil
	data.StateAction = windows.WTD_STATEACTION_CLOSE
	windows.WinVerifyTrustEx(windows.InvalidHWND, &windows.WINTRUST_ACTION_GENERIC_VERIFY_V2, data)
	return verified
}
```

I suppose that can't hurt to do early, but I think it's still going to do it itself.

@dblohm7?

Reply:
@andrew-d and I chatted about this and we think we should copy that wireguard-windows code.

Reply:
(for the record: I'm fairly certain that `msiexec` doesn't do full revocation checking, so to be safe and prevent someone from being able to force a downgrade to an older/vulnerable and revoked version, we run `WinVerifyTrustEx` with the `WTD_REVOKE_WHOLECHAIN` flag, as above, to protect against that)